File: src/packages/agent_core/builtin_tools/web_search/provider.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True, slots=True)
class WebSearchResult:
    title: str
    url: str
    snippet: str

    def to_json(self) -> dict[str, str]:
        return {"title": self.title, "url": self.url, "snippet": self.snippet}


class WebSearchProvider(Protocol):
    async def search(self, *, query: str, max_results: int) -> list[WebSearchResult]: ...


__all__ = ["WebSearchProvider", "WebSearchResult"]

File: src/services/worker_go/internal/tools/builtin/web_search/provider.go
package websearch

import "context"

type Result struct {
	Title   string
	URL     string
	Snippet string
}

func (r Result) ToJSON() map[string]any {
	payload := map[string]any{
		"title": r.Title,
		"url":   r.URL,
	}
	if r.Snippet != "" {
		payload["snippet"] = r.Snippet
	}
	return payload
}

type Provider interface {
	Search(ctx context.Context, query string, maxResults int) ([]Result, error)
}

type HttpError struct {
	StatusCode int
}

func (e HttpError) Error() string {
	return "http error"
}

